src/datastore/helpers.ts

```diff
@@ -1343,14 +1343,26 @@ export function newReOrgUpdatedEntities(): ReOrgUpdatedEntities {
  */
 export class PgWriteQueue {
   readonly queue: PQueue;
+  private tasks: Promise<unknown>[];
   constructor() {
     const concurrency = Math.max(1, getUintEnvOrDefault('STACKS_BLOCK_DATA_INSERT_CONCURRENCY', 4));
     this.queue = new PQueue({ concurrency, autoStart: true });
+    this.tasks = [];
   }
   enqueue(task: Parameters<PQueue['add']>[0]): void {
-    void this.queue.add(task);
+    const p = this.queue.add(task);
+    p.catch(void 0);
+    this.tasks.push(p);
   }
-  done(): Promise<void> {
-    return this.queue.onIdle();
+  async done(): Promise<void> {
+    // https://medium.com/@alkor_shikyaro/transactions-and-promises-in-node-js-ca5a3aeb6b74
```

Comment:
I don't believe the problem described in this post applies to the tasks handled by this queue for a couple of reasons:
1. `PQueue`'s `onIdle` [already behaves like `allSettled`](https://www.npmjs.com/package/p-queue#onidle)
    > Returns a promise that settles when the queue becomes empty, and all promises have completed
1. The API only enqueues work to this queue inside of a pre-existing SQL transaction (all the block processing code is contained in a single `sqlWriteTransaction`). If there's ever an exception thrown inside, the transaction aborts and rolls back all the pending writes. We've done extensive tests on this and it's worked correctly for a long time.

Reply:
1. That's what I thought, but it was wrong; here's a demonstration: https://replit.com/@bestmike007/TroubledEarnestSandboxes#index.ts
2. `onIdle` does not throw an exception, it simply waits all promises to complete.

What I'm seeing is, there was an unhandledRejection and then the process restarted.